### ISSUE TITLE
Add retry suspend function

### ISSUE THREAD

**Issue by enginebai** [enginebai/AndroidBase#14]
```kotlin
// retry with exponential backoff
// inspired by https://stackoverflow.com/questions/46872242/how-to-exponential-backoff-retry-on-kotlin-coroutines
private suspend fun <T> retry(
    times: Int,
    initialDelayMillis: Long = 100,
    maxDelayMillis: Long = 1000,
    factor: Double = 2.0,
    block: suspend () -> T
): T {
    var currentDelay = initialDelayMillis
    repeat(times) {
        try {
            return block()
        } catch (exception: Exception) {
            Timber.e(exception)
        }
        delay(currentDelay)
        currentDelay = (currentDelay * factor).toLong().coerceAtMost(maxDelayMillis)
    }
    return block() // last attempt
}
```

**Comment by enginebai** [comment]
Implemented at 2ac6d1efd1c3652817d2272a4362e09e66717ea2

**Comment by enginebai** [comment]
Implemented at 494acbc
